### ISSUE TITLE
Framework handling non-CF-compliant data

### ISSUE THREAD

**Issue by yihungkuo** [NOAA-GFDL/MDTF-diagnostics#158]
@jkrasting, @tsjackson-noaa, @wrongkindofdoctor, @aradhakrishnanGFDL, @bitterbark: 
Posting this on behave of @patrickorenstein who's using the MDTF package, primarily interested in running the Convective Transition POD with output from a single-column model (SCM). This issue is related to issue #154 opened by @ahmedfiaz earlier.

When trying to run the package with the SCM model output, @patrickorenstein experienced a series of errors because the NetCDF files are not fully CF-compliant, e.g., missing units in the metadata. The message printed out by the framework in his most recent attempt to run the framework indicated one of the problems was (full message attached at the end): 

```
AttributeError: 'DataArray' object has no attribute 'axis'
```
No idea about how to solve this one yet. Will need a solution for this for @patrickorenstein soon.

A more general issue. As I remember, in the earlier versions of the package, the framework would read in data without any problems as long as the directory path, file and variable names were correctly specified. Since then, @tsjackson-noaa has added mechanisms into the framework so that it can detect the units, time format, etc., as planed. Now the framework cannot work with data that are not fully CF-compliant. Not sure whether this should be considered a feature or a bug. Note that there are NetCDF files from the official CMIP6 archive that are not CF-compliant (as @ahmedfiaz's example demonstrated). Also, people may want to run the framework with their own model output but making NetCDF files fully CF-compliant is non-trivial (@patrickorenstein has spent hours/days trying without sucess by far). I would like to suggest the lead team rethinking whether to allow the framework to accept non-CF-compliant data (e.g., considering adding a switch to turn off some of the checking steps before calling PODs).

Full message printed out by the framework:
```
$ ./mdtf -f src/default_tests_twpice.jsonc
=== Starting /Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/mdtf_framework.py

PACKAGE SETTINGS:
  case_list(0): 
    CASENAME: SCM_TWPICE
    model: TWPICE
    convention: TWPICE
    FIRSTYR: 2006
    LASTYR: 2006
    pod_list: 
      convective_transition_diag
    experiment: ""
  paths: 
    CODE_ROOT: /Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics
    OBS_DATA_ROOT: /Users/patrickorenstein/Documents/mdtf/inputdata/obs_data
    MODEL_DATA_ROOT: /Users/patrickorenstein/Documents/mdtf/inputdata/model
    WORKING_DIR: /Users/patrickorenstein/Documents/mdtf/wkdir
    OUTPUT_DIR: /Users/patrickorenstein/Documents/mdtf/wkdir
    conda_root: /Users/patrickorenstein/anaconda2
    conda_env_root: ""
    TEMP_DIR_ROOT: /Users/patrickorenstein/Documents/mdtf/wkdir
  settings: 
    convention: CMIP
    data_manager: Local_File
    environment_manager: Conda
    runtime_manager: subprocess
    file_transfer_timeout: 300
    output_manager: html
    make_variab_tar: true
  env_vars: 
    RGB: /Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/shared/rgb
    MPLBACKEND: Agg
### Framework: initialize SCM_TWPICE
WARNING: No non-console-based loggers configured.
### Framework: request data for SCM_TWPICE
Querying <pr (=precipitation_flux)>
Querying <prw (=atmosphere_mass_content_of_water_vapor)>
Querying <tave (=mass_weighted_column_average_temperature)>
No data found for <tave (=mass_weighted_column_average_temperature)>.
Querying <qsat_int (=column_integrated_saturation_humidity)>
No data found for <qsat_int (=column_integrated_saturation_humidity)>.
Request for <#4.convective_transition_diag:tave (=tave) @ 1hr> failed; finding alternate vars.
Request for <#5.convective_transition_diag:qsat_int (=qsat_int) @ 1hr> failed; finding alternate vars.
Querying <ta (=air_temperature)>
Fetching <#2.convective_transition_diag:pr (=pr) @ 1hr>
Fetching <#3.convective_transition_diag:prw (=prw) @ 1hr>
Fetching <#6.convective_transition_diag:ta (=ta) @ 1hr>
Processing <#2.convective_transition_diag:pr (=pr) @ 1hr>
WARNING: No standard_name found in dataset for 'pr'; setting to 'precipitation_flux'.
WARNING: Dataset has dimension coordinate 'lat' of size 1 not identified as scalar coord.
ERROR: Caught exception processing <#2.convective_transition_diag:pr (=pr) @ 1hr> with data_key=(21,): DataPreprocessError(Data preprocessing error for data in pr: Error in read/parse data for <#2.convective_transition_diag:pr>.)
Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 570, in process
    ds = xr_parser.DatasetParser().parse(ds, var)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 712, in parse
    self.check_variable(ds, var.translation)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 660, in check_variable
    ds_axes = [c.axis for c in ds_scalars]
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 660, in <listcomp>
    ds_axes = [c.axis for c in ds_scalars]
  File "/Users/patrickorenstein/anaconda2/envs/_MDTF_base/lib/python3.7/site-packages/xarray/core/common.py", line 229, in __getattr__
    "{!r} object has no attribute {!r}".format(type(self).__name__, name)
AttributeError: 'DataArray' object has no attribute 'axis'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 671, in preprocess_data
    pod.preprocessor.process(v)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 573, in process
    f"{var.full_name}."), var) from exc
src.util.exceptions.DataPreprocessError: Data preprocessing error for data in pr: Error in read/parse data for <#2.convective_transition_diag:pr>.
Processing <#3.convective_transition_diag:prw (=prw) @ 1hr>
WARNING: No standard_name found in dataset for 'prw'; setting to 'atmosphere_mass_content_of_water_vapor'.
WARNING: Dataset has dimension coordinate 'lat' of size 1 not identified as scalar coord.
ERROR: Caught exception processing <#3.convective_transition_diag:prw (=prw) @ 1hr> with data_key=(22,): DataPreprocessError(Data preprocessing error for data in prw: Error in read/parse data for <#3.convective_transition_diag:prw>.)
Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 570, in process
    ds = xr_parser.DatasetParser().parse(ds, var)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 712, in parse
    self.check_variable(ds, var.translation)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 660, in check_variable
    ds_axes = [c.axis for c in ds_scalars]
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 660, in <listcomp>
    ds_axes = [c.axis for c in ds_scalars]
  File "/Users/patrickorenstein/anaconda2/envs/_MDTF_base/lib/python3.7/site-packages/xarray/core/common.py", line 229, in __getattr__
    "{!r} object has no attribute {!r}".format(type(self).__name__, name)
AttributeError: 'DataArray' object has no attribute 'axis'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 671, in preprocess_data
    pod.preprocessor.process(v)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 573, in process
    f"{var.full_name}."), var) from exc
src.util.exceptions.DataPreprocessError: Data preprocessing error for data in prw: Error in read/parse data for <#3.convective_transition_diag:prw>.
Processing <#6.convective_transition_diag:ta (=ta) @ 1hr>
WARNING: No standard_name found in dataset for 'ta'; setting to 'air_temperature'.
WARNING: cf_xarray fix: assuming plm is Z axis for ta
ERROR: Caught exception processing <#6.convective_transition_diag:ta (=ta) @ 1hr> with data_key=(20,): DataPreprocessError(Data preprocessing error for data in ta: Error in read/parse data for <#6.convective_transition_diag:ta>.)
Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 570, in process
    ds = xr_parser.DatasetParser().parse(ds, var)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 712, in parse
    self.check_variable(ds, var.translation)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 635, in check_variable
    raise TypeError(f"Variable {tv_name} has unexpected dimensionality: "
TypeError: Variable ta has unexpected dimensionality:  expected axes {'Y', 'X', 'T', 'Z'}, got {'Y', 'X', 'T'}.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 671, in preprocess_data
    pod.preprocessor.process(v)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 573, in process
    f"{var.full_name}."), var) from exc
src.util.exceptions.DataPreprocessError: Data preprocessing error for data in ta: Error in read/parse data for <#6.convective_transition_diag:ta>.
Request for <#2.convective_transition_diag:pr (=pr) @ 1hr> failed; finding alternate vars.
No alternates available for <#2.convective_transition_diag:pr>.
Request for <#3.convective_transition_diag:prw (=prw) @ 1hr> failed; finding alternate vars.
No alternates available for <#3.convective_transition_diag:prw>.
Request for <#6.convective_transition_diag:ta (=ta) @ 1hr> failed; finding alternate vars.
No alternates available for <#6.convective_transition_diag:ta>.
WARNING: Execution of POD convective_transition_diag couldn't be completed:
***** Caught exception #1:
Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 570, in process
    ds = xr_parser.DatasetParser().parse(ds, var)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 712, in parse
    self.check_variable(ds, var.translation)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 660, in check_variable
    ds_axes = [c.axis for c in ds_scalars]
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 660, in <listcomp>
    ds_axes = [c.axis for c in ds_scalars]
  File "/Users/patrickorenstein/anaconda2/envs/_MDTF_base/lib/python3.7/site-packages/xarray/core/common.py", line 229, in __getattr__
    "{!r} object has no attribute {!r}".format(type(self).__name__, name)
AttributeError: 'DataArray' object has no attribute 'axis'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 671, in preprocess_data
    pod.preprocessor.process(v)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 573, in process
    f"{var.full_name}."), var) from exc
src.util.exceptions.DataPreprocessError: Data preprocessing error for data in pr: Error in read/parse data for <#2.convective_transition_diag:pr>.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 489, in _check_variable
    f"since all data_keys eliminated ({dk_list})."), v)
src.util.exceptions.GenericDataSourceError: General DataSource error for data in pr: Deactivating <#2.convective_transition_diag:pr> since all data_keys eliminated ([(21,)]).

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/diagnostic.py", line 559, in update_active_vars
    self) from v.exception
src.util.exceptions.PodDataError: Requested data not available for POD 'convective_transition_diag': No alternates available for <#2.convective_transition_diag:pr (=pr) @ 1hr>.

***** Caught exception #2:
Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 570, in process
    ds = xr_parser.DatasetParser().parse(ds, var)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 712, in parse
    self.check_variable(ds, var.translation)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 660, in check_variable
    ds_axes = [c.axis for c in ds_scalars]
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 660, in <listcomp>
    ds_axes = [c.axis for c in ds_scalars]
  File "/Users/patrickorenstein/anaconda2/envs/_MDTF_base/lib/python3.7/site-packages/xarray/core/common.py", line 229, in __getattr__
    "{!r} object has no attribute {!r}".format(type(self).__name__, name)
AttributeError: 'DataArray' object has no attribute 'axis'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 671, in preprocess_data
    pod.preprocessor.process(v)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 573, in process
    f"{var.full_name}."), var) from exc
src.util.exceptions.DataPreprocessError: Data preprocessing error for data in prw: Error in read/parse data for <#3.convective_transition_diag:prw>.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 489, in _check_variable
    f"since all data_keys eliminated ({dk_list})."), v)
src.util.exceptions.GenericDataSourceError: General DataSource error for data in prw: Deactivating <#3.convective_transition_diag:prw> since all data_keys eliminated ([(22,)]).

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/diagnostic.py", line 559, in update_active_vars
    self) from v.exception
src.util.exceptions.PodDataError: Requested data not available for POD 'convective_transition_diag': No alternates available for <#3.convective_transition_diag:prw (=prw) @ 1hr>.

***** Caught exception #3:
Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 570, in process
    ds = xr_parser.DatasetParser().parse(ds, var)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 712, in parse
    self.check_variable(ds, var.translation)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/xr_parser.py", line 635, in check_variable
    raise TypeError(f"Variable {tv_name} has unexpected dimensionality: "
TypeError: Variable ta has unexpected dimensionality:  expected axes {'Y', 'X', 'T', 'Z'}, got {'Y', 'X', 'T'}.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 671, in preprocess_data
    pod.preprocessor.process(v)
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/preprocessor.py", line 573, in process
    f"{var.full_name}."), var) from exc
src.util.exceptions.DataPreprocessError: Data preprocessing error for data in ta: Error in read/parse data for <#6.convective_transition_diag:ta>.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/data_manager.py", line 489, in _check_variable
    f"since all data_keys eliminated ({dk_list})."), v)
src.util.exceptions.GenericDataSourceError: General DataSource error for data in ta: Deactivating <#6.convective_transition_diag:ta> since all data_keys eliminated ([(20,)]).

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/diagnostic.py", line 559, in update_active_vars
    self) from v.exception
src.util.exceptions.PodDataError: Requested data not available for POD 'convective_transition_diag': No alternates available for <#6.convective_transition_diag:ta (=ta) @ 1hr>.

.
### Framework: Data request for SCM_TWPICE failed, skipping execution.
SCM_TWPICE: Creating /Users/patrickorenstein/Documents/mdtf/wkdir/MDTF_SCM_TWPICE_2006_2006.v8.tar.
--------------------------------------------------------------------------------
Exiting with errors from /Users/patrickorenstein/Documents/mdtf/MDTF-diagnostics/src/core.py
Summary for SCM_TWPICE:
	The following PODs raised errors: convective_transition_diag
	Output written to /Users/patrickorenstein/Documents/mdtf/wkdir/MDTF_SCM_TWPICE_2006_2006.v8
```

**Comment by yihungkuo** [comment]
Remark: As @tsjackson-noaa has noted, the size along the latitude direction is 1. The original data is from a single-column model which outputs at a fixed location. The spatial dimension is manually extended to 4x1 (lon x lat) so the convective transition POD can work with it. If the framework or xarray parser requires the array size along any dimension to be greater than 1, P.D.O. can do that.

**Comment by wrongkindofdoctor** [comment]
@yihungkuo As noted at today's mini leads meeting, @tsjackson-noaa and I will work on a fix that will likely involve adding an option to fill in default data during pre-processing.

**Comment by tsjackson-noaa** [comment]
Closing this issue as resolved by PR #161.
